src/splinterdb.c

```diff
@@ -227,38 +229,95 @@ splinterdb_init_config(const splinterdb_config *kvs_cfg, // IN
    platform_status status;
 
+   bool we_created_heap = FALSE;
+
+   // Allocate a shared segment if so requested. For now, we hard-code
+   // the required size big enough to run most tests. Eventually this
+   // has to be calculated here based on other run-time params.
+   // (Some tests externally create the platform_heap, so we should
+   // only create one if it does not already exist.)
+   if (kvs_cfg->use_shmem && (kvs_cfg->heap_handle == NULL)) {
+      size_t shmem_size = (kvs_cfg->shmem_size ? kvs_cfg->shmem_size : 2 * GiB);
+      status            = platform_heap_create(platform_get_module_id(),
+                                    shmem_size,
+                                    TRUE,
+                                    &kvs_cfg->heap_handle,
+                                    &kvs_cfg->heap_id);
+      if (!SUCCESS(status)) {
+         platform_error_log(
+            "Shared memory creation failed. "
+            "Failed to %s SplinterDB device '%s' with specified "
+            "configuration: %s\n",
+            (open_existing ? "open existing" : "initialize"),
+            kvs_cfg->filename,
+            platform_status_to_string(status));
+         goto deinit_kvhandle;
+      }
+      we_created_heap = TRUE;
+
+      // Setup global tracing booleans for shared memory usage
+      if (kvs_cfg->trace_shmem_allocs) {
+         platform_enable_tracing_shm_allocs();
+      }
+      if (kvs_cfg->trace_shmem_frees) {
+         platform_enable_tracing_shm_frees();
+      }
+      if (kvs_cfg->trace_shmem) {
+         platform_enable_tracing_shm_frees();
+         platform_enable_tracing_shm_ops();
+      }
+   }
+
    platform_assert(kvs_out != NULL);
 
    kvs = TYPED_ZALLOC(kvs_cfg->heap_id, kvs);
    if (kvs == NULL) {
       status = STATUS_NO_MEMORY;
+      if (we_created_heap) {
+         platform_heap_destroy(&kvs_cfg->heap_handle);
+      }
       return platform_status_to_int(status);
    }
 
+   platform_heap_handle heap_handle = NULL;
+
+   // All memory allocation after this call should -ONLY- use heap handles
+   // from the handle to the running Splinter instance; i.e. 'kvs'.
    status = splinterdb_init_config(kvs_cfg, kvs);
    if (!SUCCESS(status)) {
       platform_error_log("Failed to %s SplinterDB device '%s' with specified "
                          "configuration: %s\n",
                          (open_existing ? "open existing" : "initialize"),
                          kvs_cfg->filename,
                          platform_status_to_string(status));
+      heap_handle = kvs_cfg->heap_handle;
       goto deinit_kvhandle;
    }
 
+   // All future memory allocation should come from shared memory, if so
+   // configured.
+   kvs->heap_handle = kvs_cfg->heap_handle;
+   kvs->heap_id     = kvs_cfg->heap_id;
+
+   // This allows for a simple usage where application can close and reopen
+   // Splinter and not run into seg-faults due to stale memory handles.
+   kvs_cfg->heap_handle = NULL;
```

Comment:
This is why the input `const` qualifier was dropped. There is at least one test case that wants to close & reopen Splinter. If the heap-handles off of input `kvs_cfg` are retained, then due to the way the logic is implemented on L 247 above, for the usage scenario of a close-and-reopen we will end up with an invalid `kvs_cfg->heap_handle`.

All hell breaks loose.

This rework accommodates all use-cases. If there is a better way to rework this exercise / usage, I'll be happy to hear suggestions.

Reply:
Resolved. Reworked interfaces to re-introduce const splinterdb_config *kvs_cfg.